accepted/2021/preview-features/preview-features.md

```diff
@@ -0,0 +1,420 @@
+* NuGet team
+* PowerShell team
+    - They may want to/need to do work to prevent PowerShell scripts taking
+      accidental dependency on preview features.
+
+## Design
+
+### Property in project files
+
+```xml
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net6.0</TargetFramework>
+    <EnablePreviewFeatures>True</EnablePreviewFeatures>
+  </PropertyGroup>
+
+</Project>
+```
+
+### Assembly info generation
+
+When we generate the `AssemblyInfo.cs`, we'll generate an assembly level
+attribute (the attribute and the meaning is discussed [later in this
+document](#api-analyzer)).
+
+For example, this project information:
+
+```xml
+<Project>
+
+  <PropertyGroup>
+    <TargetFramework>net6.0</TargetFramework>
+    <EnablePreviewFeatures>True</EnablePreviewFeatures>
+  </PropertyGroup>
+
+</Project>
+```
+
+Would result in this assembly-level attribute:
+
+```C#
+[assembly: RequiresPreviewFeatures]
+```
+
+This means you can look at a binary and tell whether it was built using preview
+features. This can be used for auditing and for detecting transitive
+dependencies ([discussed later](#api-analyzer)).
+
+### runtime.json
+
+Whether or not `EnablePreviewFeatures` was set to true for the application
+should be written into the `runtime.json` file such that the host/runtime can
+use it to make decisions. Alternatively, we could say this is encoded by the
+fact that the assembly the host considers the entry point has the
+`RequiresPreviewFeatures` assembly level attribute.
+
+### Compilation context
+
+When `EnablePreviewFeatures` is true, the `LangVersion` property should be set
+to `Preview`. This avoids customers having to turn on preview features for the
+language separately.
+
+In addition, the property `EnablePreviewFeatures` should be passed to the
+compilation context (akin to how `TargetFramework` is passed in today). An
+analyzer will use that to block use of APIs that are marked as preview as basing
+this off the language preview mode feels wrong.
+
+***Note:*** F# doesn't have this capability today.
+
+### API Analyzer
+
+We'll need an attribute to record whether or not the project was
+built with preview features turned on:
+
+```C#
+namespace System.Runtime.Versioning
+{
+    [AttributeUsage(AttributeTargets.Assembly |
+                    AttributeTargets.Module |
+                    AttributeTargets.Class |
+                    AttributeTargets.Struct |
+                    AttributeTargets.Enum |
+                    AttributeTargets.Constructor |
+                    AttributeTargets.Method |
+                    AttributeTargets.Property |
+                    AttributeTargets.Field |
+                    AttributeTargets.Event, AllowMultiple = true, Inherited = false)]
+    public sealed class RequiresPreviewFeaturesAttribute : Attribute
+    {
+        public RequiresPreviewFeaturesAttribute();
+    }
+}
+```
+
+We'll ship a built-in analyzer that will report diagnostics when an API is being
```

Comment:
Are we planning to use the preview mode to change the behavior of an existing API (maybe to fix a bug or to extend the range of functionality of a method)? 
Then it would not be possible to mark the API with `[RequiresPreviewFeatures]`. We'd potentially need a runtime check in the library.

Reply:
I don't think so. We'd simply make the change in a preview build of the platform and/or add a compat switch to get the old behavior back if deemed necessary.